src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,929 @@
+    // is a 0 that should be a 1. Obviously, since the bit at that position in
+    // ones is 1, the same bit in zeros is 0. Which means this is the value of
+    // i we are looking for.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 1 0 0 1 0
+    //   zeros = 0 0 1 0 0 1 0 0
+    //    ones = 0 1 0 0 1 0 1 0
+    //   1-vio = 0 1 0 0 1 0 0 0
+    //   0-vio = 0 0 0 0 0 0 0 0
+    // Since the result must have the 2nd bit set, it must be at least:
+    //           1 1 0 0 0 0 0 0
+    // This value must satisfy zeros, because all bits before the 2nd bit have
+    // already satisfied zeros, and all bits after the 2nd bit are all 0 now.
+    // Just OR this value with ones to obtain the final result.
+
+    // first_violation is the position of the violation counting from the
+    // highest bit down (0-based), since i == 2, first_difference == 1
+    juint first_violation = count_leading_zeros(one_violation); // 1
+    //           1 0 0 0 0 0 0 0
+    constexpr U highest_bit = (std::numeric_limits<U>::max() >> 1) + U(1);
+    //           0 1 0 0 0 0 0 0
+    U alignment = highest_bit >> first_violation;
+    // This is the first value which have the violated bit being 1, which means
+    // that the result should not be smaller than this
+    //           1 1 0 0 0 0 0 0
+    U new_lo = (lo & -alignment) + alignment;
+    //           1 1 0 0 1 0 1 0
+    new_lo |= bits._ones;
+    assert(lo < new_lo, "this case cannot overflow");
+    return new_lo;
+  } else {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 1 that should be a 0. Trace backward to find i which is the last
+    // bit that is 0 in both lo and zeros.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 0 1 1 1 0
+    //   zeros = 0 0 0 1 0 1 0 0
+    //    ones = 1 0 0 0 0 0 1 1
+    //   1-vio = 0 0 0 0 0 0 0 1
+    //   0-vio = 0 0 0 0 0 1 0 0
+    // The first violation is the 6th bit, which should be 0. The 5th cannot be
+    // the first different bit we are looking for, because it is already 1, the
+    // 4th bit also cannot be, because it must be 0. As a result, the first
+    // different bit between the result and lo must be the 3rd bit. As a result,
+    // the result must not be smaller than:
+    //           1 0 1 0 0 0 0 0
+    // This one satisfies zeros so we can use the logic in the previous case to
+    // obtain our final result, which is:
+    //           1 0 1 0 0 0 1 1
+
+    juint first_violation = count_leading_zeros(zero_violation);
+    // This mask out all bits from the first violation
+    //           1 1 1 1 1 0 0 0
+    U find_mask = ~(std::numeric_limits<U>::max() >> first_violation);
+    //           1 0 0 1 1 1 1 0
+    U either = lo | bits._zeros;
+    // i is the last bit being 0 in either that stands before the first
+    // violation, which is the last set bit of tmp
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
```

Comment:
Did I understand that right: `tmp` is the bits that we cannot flip? Or is it the ones we can flip? A better name would be appreciated :)
Same for `either`. worst case you call it `lo_or_zeros`... but that's not great either.

I'm not fully seeing through the logic here yet, so I struggle to make good suggestions.

Reply:
We can say that `tmp` is all the bits we can flip, although I think that is too ambiguous, what does "can" mean here. It is better to think of it as all the bits that are not lower than `first_violation` and are 0 in both `lo` and `zeros`.

Reply:
Hmm ok. Well it would still be helpful to at least have some kind of intuition, and "name" for it. Are all these bits the candidates for `alignment`, of we must pick the one most to the right, i.e. so that we get the smallest `alignment` value?
And then why is it these bits, and not any others? An argument would be good here.

Reply:
Hmmm these bits are the ones that are not lower than `first_violation` and are 0 in both `lo` and `zeros`. This description aligns perfectly with the conclusion in the formality section above. So what we are finding is the last such bit here. I say that in the following computation.